src/inter_dc/reqrouter/inter_dc_query_req.erl

```diff
@@ -0,0 +1,248 @@
+    {ok, #state{sockets = dict:new(), req_id = 1}}.
+
+%% Handle the instruction to add a new DC.
+handle_call({add_dc, DCID, LogReaders}, _From, State = #state{ sockets = OldDcPartitionDict }) ->
+    %% delete the dc if already added
+    InitialDcPartitionDict = del_dc(DCID, OldDcPartitionDict),
+
+    %% add every DC-Partition pair to dict
+    AddPartitionsToDict = fun(Partition, {CurrentDict, Socket}) ->
+            Key = {DCID, Partition},
+            %% assert that DC was really deleted
+            error = dict:find(Key, CurrentDict),
+            NewDict = dict:store(Key, Socket, CurrentDict),
+            {NewDict, Socket}
+        end,
+
+    %% for each log reader add every {DCID, Partition} tuple to the dict
+    AddLogReaders = fun({Partitions, AddressList}, CurrentDict) ->
+            {ok, Socket} = connect_to_node(AddressList),
+            {ResultDict, Socket} = lists:foldl(AddPartitionsToDict, {CurrentDict, Socket}, Partitions),
+            ResultDict
+        end,
+
+    ResultDcPartitionDict = lists:foldl(AddLogReaders, InitialDcPartitionDict, LogReaders),
+    {reply, ok, State#state{sockets = ResultDcPartitionDict}};
+
+%% Remove a DC
+handle_call({del_dc, DCID}, _From, State = #state{ sockets = Dict}) ->
+    NewDict = del_dc(DCID, Dict),
+    {reply, ok, State#state{sockets = NewDict}};
+
+%% Handle an instruction to ask a remote DC.
+handle_call({any_request, RequestType, {DCID, Partition}, BinaryRequest, Func}, _From, State=#state{req_id=ReqId}) ->
+    ?LOG_NOTICE("Trying to perform request"),
+    case dict:find({DCID, Partition}, State#state.sockets) of
+        {ok, Socket} ->
+            ?LOG_DEBUG("Request and receive message async to ~p ~p (~p)", [DCID, Partition, Socket]),
+
+            %% prepare message
+            VersionBinary = ?MESSAGE_VERSION,
+            ReqIdBinary = inter_dc_txn:req_id_to_bin(ReqId),
+            FullRequest = <<VersionBinary/binary, ReqIdBinary/binary, RequestType, BinaryRequest/binary>>,
+
+            %% TODO the old variant (erlzmq) sent this request and receive pair in a different process async.
+            %%      That would pile up requests in the background, as the send and recv order is strictly
+            %%      needed for req sockets.
+            %%      It would make sense to open more sockets concurrently
+            %%      and dispatch these send/recv requests (e.g. as much as ?INTER_DC_QUERY_CONCURRENCY)
+            ?LOG_DEBUG("Send and receive"),
+            ok = chumak:send(Socket, FullRequest),
+            {ok, BinaryMsg} = chumak:recv(Socket),
+
+            ?LOG_DEBUG("Process and apply function"),
+            <<_ReqIdBinary:?REQUEST_ID_BYTE_LENGTH/binary, RestMsg/binary>> = inter_dc_utils:check_message_version(BinaryMsg),
+            case RestMsg of
+                <<RequestType, RestBinary/binary>> -> Func(RestBinary);
+                Other -> ?LOG_ERROR("Received unknown reply: ~p", [Other])
+            end,
```

Comment:
If I understand this correctly, that would be a significant change in the behavior compared to the previous version.
Now the caller would have to wait until a request is completed which is not the right behavior I think.
All processes that send requests (e.g. the bcounter_mgr) would be blocked if the connection is slow or not available.